clients/adapter-kafka/src/main/java/org/eclipse/hono/adapter/client/command/kafka/KafkaBasedCommand.java

```diff
@@ -0,0 +1,285 @@
+import java.util.Objects;
+import java.util.Optional;
+import java.util.StringJoiner;
+
+import org.eclipse.hono.adapter.client.command.Command;
+import org.eclipse.hono.adapter.client.command.Commands;
+import org.eclipse.hono.kafka.client.HonoTopic;
+import org.eclipse.hono.kafka.util.RecordUtil;
+import org.eclipse.hono.tracing.TracingHelper;
+import org.eclipse.hono.util.MessageHelper;
+import org.eclipse.hono.util.Strings;
+
+import io.opentracing.Span;
+import io.opentracing.log.Fields;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.kafka.client.consumer.KafkaConsumerRecord;
+
+/**
+ * A command used in a Kafka based client.
+ *
+ */
+public final class KafkaBasedCommand implements Command {
+
+    /**
+     * If present, the command is invalid.
+     */
+    private final Optional<String> validationError;
+    private final KafkaConsumerRecord<String, Buffer> record;
+    private final String tenantId;
+    private final String deviceId;
+    private final String originalDeviceId;
+    private final String correlationId;
+    private final String subject;
+    private final String contentType;
+    private final String requestId;
+
+    private KafkaBasedCommand(
+            final Optional<String> validationError,
+            final KafkaConsumerRecord<String, Buffer> commandRecord,
+            final String tenantId,
+            final String deviceId,
+            final String originalDeviceId,
+            final String correlationId,
+            final String subject,
+            final String contentType) {
+
+        this.validationError = validationError;
+        this.record = commandRecord;
+        this.tenantId = tenantId;
+        this.deviceId = deviceId;
+        this.originalDeviceId = originalDeviceId;
+        this.correlationId = correlationId;
+        this.subject = subject;
+        this.contentType = contentType;
+        this.requestId = Commands.getRequestId(correlationId);
+    }
+
+    /**
+     * Creates a command from a Kafka consumer record.
+     * <p>
+     * The record is expected to contain
+     * <ul>
+     * <li>a <em>topic</em> of the format <em>hono.command.${tenant_id}</em></li>
+     * <li>a <em>key</em> with the command target device id, matching the value of the <em>device_id</em> header</li>
+     * <li>a <em>subject</em> header</li>
+     * <li>a non-empty <em>correlation-id</em> header if a response is expected for the command.</li>
+     * </ul>
+     * <p>
+     * If any of the requirements above are not met, then the returned command's {@link #isValid()}
+     * method will return {@code false}.
+     *
+     * @param record The record containing the command.
+     * @param deviceId The identifier of the device that the command will be sent to. If the command has been mapped
+     *                 to a gateway, this id is the gateway id and the original command target device is given in
+     *                 the <em>device_id</em> header and the key of the record.
```

Comment:
FMPOV this is very confusing. Why isn't all the necessary information contained in the headers? Why do we need to pass in the device ID as a separate parameter? Can't we simply read the target device ID from the *device_id* header and, if the device is connected via a gateway, read the gateway ID from e.g. the *gateway_id* header that has been added by the Command Router?

This is also very confusing in the `Command` interface as well. FMPOV the *getDeviceId()* method should *always* return the identifier of the device that is the target of the command. We should then replace *getOriginalDeviceId* with *getGatewayId* which returns the gateway ID if the command is routed via a gateway. WDYT?

Reply:
Yes, such a change is something I already had on my TODO list. As there are quite many files involved in that change, I would rather do this in a separate PR, created as a followup of this one.

>Why do we need to pass in the device ID as a separate parameter?

Because there is one scenario, where the gateway id cannot be deduced from the message/record (i.e. the `via` property/header set by the command router). That's when a gateway subscribes for commands of a specific device. The gateway information is only available at the target protocol adapter in this case.
But I want to remove the device id parameter from the factory method and either set the `via` property/header beforehand or introduce a `setGatewayId()` setter for that.